sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -953,15 +952,16 @@ case class Pmod(
           // when we reach here, failOnError must bet true.
           throw QueryExecutionErrors.divideByZeroError(queryContext)
         }
-        input1 match {
-          case i: Integer => pmod(i, input2.asInstanceOf[java.lang.Integer])
-          case l: Long => pmod(l, input2.asInstanceOf[java.lang.Long])
-          case s: Short => pmod(s, input2.asInstanceOf[java.lang.Short])
-          case b: Byte => pmod(b, input2.asInstanceOf[java.lang.Byte])
-          case f: Float => pmod(f, input2.asInstanceOf[java.lang.Float])
-          case d: Double => pmod(d, input2.asInstanceOf[java.lang.Double])
-          case d: Decimal => checkOverflow(
-            pmod(d, input2.asInstanceOf[Decimal]), dataType.asInstanceOf[DecimalType])
+
+        dataType match {
```

Comment:
we should cache it somewhere, otherwise it's very expensive to call it per row.

Reply:
e.g. maybe we can call `dataType` only once and create a `(Any, Any) => Any` lambda and use it here.